crates/stages/src/sets.rs

```diff
@@ -0,0 +1,181 @@
+where
+    DB: Database,
+    H: HeaderDownloader + 'static,
+    B: BodyDownloader + 'static,
+{
+    fn build(self) -> StageSetBuilder<DB> {
+        self.online.build().add_set(OfflineStages)
+    }
+}
+
+/// A set containing all stages that require network access by default.
+///
+/// These stages *can* be run without network access if the specified downloaders are
+/// themselves offline.
+#[derive(Debug)]
+pub struct OnlineStages<H, B> {
+    /// The consensus engine used to validate incoming data.
+    consensus: Arc<dyn Consensus>,
+    /// The block header downloader
+    header_downloader: H,
+    /// The block body downloader
+    body_downloader: B,
+}
+
+impl<H, B> OnlineStages<H, B> {
+    /// Create a new set of online stages with default values.
+    pub fn new(consensus: Arc<dyn Consensus>, header_downloader: H, body_downloader: B) -> Self {
+        Self { consensus, header_downloader, body_downloader }
+    }
+}
+
+impl<DB, H, B> StageSet<DB> for OnlineStages<H, B>
+where
+    DB: Database,
+    H: HeaderDownloader + 'static,
+    B: BodyDownloader + 'static,
+{
+    fn build(self) -> StageSetBuilder<DB> {
+        StageSetBuilder::default()
+            .add_stage(HeaderStage::new(self.header_downloader, self.consensus.clone()))
+            .add_stage(TotalDifficultyStage::default())
+            .add_stage(BodyStage { downloader: self.body_downloader, consensus: self.consensus })
+            .add_stage(TransactionLookupStage::default())
+    }
+}
+
+/// A set containing all stages that do not require network access.
+///
+/// A combination of (in order)
+///
+/// - [`ExecutionStages`]
+/// - [`HashingStages`]
+#[derive(Debug)]
+pub struct OfflineStages;
+
+impl<DB: Database> StageSet<DB> for OfflineStages {
+    fn build(self) -> StageSetBuilder<DB> {
+        ExecutionStages::default().build().add_set(HashingStages).add_set(HistoryIndexingStages)
+    }
+}
+
+/// A set containing all stages that are required to execute pre-existing block data.
+#[derive(Debug)]
+pub struct ExecutionStages {
+    /// The chain specification to use for execution.
+    chain_spec: ChainSpec,
+}
+
+impl Default for ExecutionStages {
+    fn default() -> Self {
+        Self { chain_spec: reth_primitives::MAINNET.clone() }
+    }
+}
+
+impl ExecutionStages {
+    /// Create a new set of execution stages with default values.
+    pub fn new(chain_spec: ChainSpec) -> Self {
+        Self { chain_spec }
+    }
+}
+
+impl<DB: Database> StageSet<DB> for ExecutionStages {
+    fn build(self) -> StageSetBuilder<DB> {
+        StageSetBuilder::default()
+            .add_stage(SenderRecoveryStage::default())
+            .add_stage(ExecutionStage { chain_spec: self.chain_spec, ..Default::default() })
+    }
+}
+
+/// A set containing all stages that hash account state.
+#[derive(Debug)]
+pub struct HashingStages;
+
+impl<DB: Database> StageSet<DB> for HashingStages {
+    fn build(self) -> StageSetBuilder<DB> {
```

Comment:
giga nit: Maybe call `builder` given this doesn't actually build and it returns a builder?

```suggestion
    fn builder(self) -> StageSetBuilder<DB> {
```
```suggestion
    fn build(self) -> StageSetBuilder<DB> {
```